sympy/geometry/polygon.py

```diff
@@ -1051,17 +1054,49 @@ def __eq__(self, o):
                     return True
         return False
 
-    def __hash__(self):
-
-        # getKey function helps in sorting Point objects.
-        def getKey(custom):
-            return custom.args
-
-        # Temporary Polygon is created to get same value of hash for Polygons
-        # with different sequence of vertices. Sorted tuple of vertices is
-        # used to initialise this polygon.
-        temp = Polygon( *sorted( self.args, key = getKey))
-        return super(Polygon, temp).__hash__()
+    def _hashable_content(self):
+        
+        def rotate_string( a, offset):
+            return a[offset:] + a[:offset]
+        
+        def convert_string( point_list):
+            '''
+            Converts list of Point objects to String containg co-ordinates
+            '''
+            S = ''
+            for point in point_list:
+                S += ''.join(map(str, point))
+            return S
+        
+        # Implementation of Booth's Algorithm to find Lexicographically minimal string rotation
+        # See https://en.wikipedia.org/wiki/Lexicographically_minimal_string_rotation
+        def least_rotation(S):
+            S += S      
+            f = [-1] * len(S)     # Failure function
+            k = 0       # Least rotation of string found so far
+            
+            for j in range(1,len(S)):
+                sj = S[j]
+                i = f[j-k-1]
+                while i != -1 and sj != S[k+i+1]:
+                    if sj < S[k+i+1]:
+                        k = j-i-1
+                    i = f[i]
+                if sj != S[k+i+1]:
+                    if sj < S[k]: 
+                        k = j
+                    f[j-k] = -1
+                else:
+                    f[j-k] = i+1
+            return k
+        # find minimal rotation of both clockwise and anticlockwise list of Points
```

Comment:
Since the polygons are directed, they should not compare the same if they have different traversal order, so I think you should not return the smaller of either direction.

Reply:
Polygons with cyclic or reverse cyclic order are considered equal in geometry module.
So smallest rotation in one direction wont give same string if one is in cyclic order and other is reverse cyclic.
If only one direction rotation is used,
   (0, 0), (1, 0), (1, 1) and (1, 1), (1, 0), (0, 0))
  Result will be
  (0, 0), (1, 0), (1, 1)) for first list
  (0, 0), (1, 1), (1, 0) for second list.

Reply:
> Polygons with cyclic or reverse cyclic order are considered equal in geometry module

I'm not sure they should be since, as is asked on the issue page, "should triangles with opposite areas due to point ordering really be comparing equal?" Perhaps with "equals" they can be the same but with `==` they will not be? Or maybe a "canonical" method could be used to always orient them to have the smallest cyclical ordering.

Reply:
I am in favour of it. It may cause problem in existing code. @asmeurer What do you think?